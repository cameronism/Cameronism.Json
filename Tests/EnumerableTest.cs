using System;
using System.Collections.Generic;
using System.Linq;
using System.Linq.Expressions;
using System.Reflection;
using System.Text;
using System.Threading.Tasks;
using UnsafeJson;
using Xunit;

namespace Tests
{
	public class EnumerableTest
	{
		class ExplicitEnumerableBadCount : IEnumerable<bool>
		{
			IEnumerator<bool> IEnumerable<bool>.GetEnumerator() { throw new NotImplementedException(); } 
			System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator() { throw new NotImplementedException(); } 
			/// <summary>
			/// This should not be picked up since neither ICollection<> or IReadOnlyCollection are implemented
			/// </summary>
			public int Count
			{
				get
				{
					return 0;
				}
			}
		}

		class ExplicitReadOnlyCollection : IReadOnlyCollection<Guid>
		{
			int IReadOnlyCollection<Guid>.Count { get { throw new NotImplementedException(); } } 
			IEnumerator<Guid> IEnumerable<Guid>.GetEnumerator() { throw new NotImplementedException(); } 
			System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator() { throw new NotImplementedException(); }
		}
		class ExplicitReadOnlyCollectionImplicitCount : IReadOnlyCollection<byte>
		{
			public int Count { get { throw new NotImplementedException(); } }
			IEnumerator<byte> IEnumerable<byte>.GetEnumerator() { throw new NotImplementedException(); }
			System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator() { throw new NotImplementedException(); }
		}
		class ExplicitReadOnlyCollectionPublicEnumerator : IReadOnlyCollection<sbyte>
		{
			public struct Enumerator
			{
				public bool MoveNext() { return false; }
				public sbyte Current { get { return 0; } }
			}

			public Enumerator GetEnumerator() { return new Enumerator(); }
			int IReadOnlyCollection<sbyte>.Count { get { throw new NotImplementedException(); } } 
			IEnumerator<sbyte> IEnumerable<sbyte>.GetEnumerator() { throw new NotImplementedException(); } 
			System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator() { throw new NotImplementedException(); }
		}



		[Fact]
		public void FindMethods()
		{
			var sb = new StringBuilder();

			foreach (var type in new[]
			{
				typeof(List<int>),
				typeof(IEnumerable<int>),
				typeof(ExplicitEnumerableBadCount),
				typeof(ExplicitReadOnlyCollection),
				typeof(ExplicitReadOnlyCollectionImplicitCount),
				typeof(ExplicitReadOnlyCollectionPublicEnumerator),
				typeof(ICollection<short>),
				typeof(IReadOnlyCollection<ushort>),
				typeof(IReadOnlyList<long>),
				typeof(IList<ulong>),
			})
			{
				var methods = EnumerableInfo.FindMethods(type);
				sb.AppendLine("# " + SchemaTest.HumanName(type));
				DescribeMethod(sb, () => methods.GetEnumerator);
				DescribeMethod(sb, () => methods.MoveNext);
				DescribeMethod(sb, () => methods.get_Current);
				DescribeMethod(sb, () => methods.get_Count);
				DescribeMethod(sb, () => methods.Dispose);

				sb.AppendLine();
			}

			ApprovalTests.Approvals.Verify(sb.ToString());
		}

		static void DescribeMethod(StringBuilder sb, Expression<Func<MethodInfo>> fmi)
		{
			var prop = fmi.Body as MemberExpression;
			sb.AppendLine("## " + prop.Member.Name);

			var mi = fmi.Compile().Invoke();
			if (mi == null)
			{
				sb.AppendLine("`null`");
			}
			else
			{
				sb.AppendLine("Return Type: " + SchemaTest.HumanName(mi.ReturnType));
				sb.AppendLine("IsVirtual: " + mi.IsVirtual);
				sb.AppendLine("Declaring Type: " + SchemaTest.HumanName(mi.DeclaringType));
				foreach (var arg in mi.GetParameters())
				{
					sb.AppendLine(String.Format("{0}: {1}", arg.Name, SchemaTest.HumanName(arg.ParameterType)));
				}
			}

			sb.AppendLine();
		}



	}
}
