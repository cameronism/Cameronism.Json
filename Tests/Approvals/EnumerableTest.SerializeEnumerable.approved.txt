

# System.Collections.Generic.IEnumerable<System.Int32>


## Newtonsoft 0


0000   6e 75 6c 6c                                         null

## UnsafeJson 0


0000   6e 75 6c 6c                                         null

### Equal: True

## Newtonsoft 1


0000   5b 5d                                               []

## UnsafeJson 1


0000   5b 5d                                               []

### Equal: True

## Newtonsoft 2


0000   5b 31 5d                                            [1]

## UnsafeJson 2


0000   5b 31 5d                                            [1]

### Equal: True

## Newtonsoft 3


0000   5b 31 2c 32 5d                                      [1,2]

## UnsafeJson 3


0000   5b 31 2c 32 5d                                      [1,2]

### Equal: True

## Newtonsoft 4


0000   5b 5d                                               []

## UnsafeJson 4


0000   5b 5d                                               []

### Equal: True

## Newtonsoft 5


0000   5b 32 5d                                            [2]

## UnsafeJson 5


0000   5b 32 5d                                            [2]

### Equal: True

## Newtonsoft 6


0000   5b 32 2c 34 5d                                      [2,4]

## UnsafeJson 6


0000   5b 32 2c 34 5d                                      [2,4]

### Equal: True

## Newtonsoft 7


0000   5b 5d                                               []

## UnsafeJson 7


0000   5b 5d                                               []

### Equal: True

## Newtonsoft 8


0000   5b 5d                                               []

## UnsafeJson 8


0000   5b 5d                                               []

### Equal: True

## Newtonsoft 9


0000   5b 33 5d                                            [3]

## UnsafeJson 9


0000   5b 33 5d                                            [3]

### Equal: True

## Newtonsoft 10


0000   5b 33 2c 33 5d                                      [3,3]

## UnsafeJson 10


0000   5b 33 2c 33 5d                                      [3,3]

### Equal: True

## Newtonsoft 11


0000   6e 75 6c 6c                                         null

## UnsafeJson 11


0000   6e 75 6c 6c                                         null

### Equal: True